ggshield/utils/git_shell.py

```diff
@@ -234,9 +234,31 @@ def git_ls(wd: Optional[Union[str, Path]] = None) -> List[str]:
     return git(cmd, timeout=600, cwd=wd).split("\n")
 
 
+def _has_submodules(wd: Union[str, Path]) -> bool:
+    cmd_submodules_status = ["-C", wd, "submodule"]
+    submodule_status = git(cmd_submodules_status, timeout=600)
+    return bool(submodule_status)
+
+
 def git_ls_unstaged(wd: Optional[Union[str, Path]] = None) -> List[str]:
-    cmd = ["ls-files", "--other", "--exclude-standard"]
-    return git(cmd, timeout=600, cwd=wd).split("\n")
+    # git command to get list of unstaged files in repo
+    cmd = ["ls-files", "--others", "--exclude-standard"]
+    unstaged_files = git(cmd, timeout=600, cwd=wd).split("\n")
+
+    if wd is None:
+        wd = Path(".")
+    if _has_submodules(wd):
+        # --recurse-submodules is not compatible with --others so we need to
+        # get the list of unstaged files in each submodule
+        submodule_path = "s~^~$sm_path/~"
+        submodule_files = (
+            f"git ls-files -o --exclude-standard --full-name | sed {submodule_path}"
```

Comment:
This is not going to work on Windows, where `sed` is not available. Using the shell like this can also be dangerous (though it looks rather safe here).

I suggest parsing the output of `git submodule status` to get the path to all submodules and recalling `git_ls_unstaged` recursively on the new paths.

This can remove the need for `_has_submodules()`, which mostly does this. The code would be something like this:

```python
def git_ls_unstaged():
    # call ls-files
    for path in get_submodule_paths():
        # call git_ls_unstaged(path)
```

Reply:
I just had to handle the path of the files to ensure they will be found during the scanning process.